transport/http2_server.go

```diff
@@ -231,18 +231,21 @@ func newHTTP2Server(conn net.Conn, config *ServerConfig) (_ ServerTransport, err
 	// Check the validity of client preface.
 	preface := make([]byte, len(clientPreface))
 	if _, err := io.ReadFull(t.conn, preface); err != nil {
```

Comment:
Shouldn't we do the preface checks before we even create the `http2Server`?

How about this:

1. check preface
2. create http2Server
3. `defer func () { if err != nil { t.Close() } }()`

so we protect ourselves from this type of problem in the future.

Reply:
Added a defer close, but kept the order of preface and create transport.